### ISSUE TITLE
Improve Piano Roll

### ISSUE THREAD

**Issue by FramePerfection** [Mupen64-Rewrite/SM64LuaRedux#26]
This issue is a direct followup to #22, which introduced a basic `PianoRoll` view that can be used to more efficiently author inputs across a medium range of frames (at most a few hundred).

While this feature is usable in its current state, it is by no means perfect - as such, this issue shall serve as a hub for improvement ideas and documenting bugs related to (and/or caused by) the feature.

Specifically, issues regarding the Piano Roll that can be pinpointed more precisely should have their own... well, GitHub issue, and have _this_ issue point to _them_, rather than having discussion for those details here.

The following checklists are not comprehensive, and individual items may (and will) be extracted into separate, more concrete, issues.

# Proposed enhancements
- [ ] The semantics of the "middle column" between the mini joysticks (potentially including the movement mode indicator) and the button columns should be improved. Currently they only show the dyaw value and straining direction for frames where this value applies - more differentiated information could and should be shown here. Additionally, this column may be used to provide "quick actions" such as changing the straining direction or the exact dyaw value without needing to select the respective frame "for editing".

- [ ] Sparse joystick selections should be made available for at least the basic joystick capabilities (i.e. excluding features like arctan straining)

- [ ] Copy-paste, duplicate, and bulk delete functionalities should be provided - including the option to toggle between overwriting and inserting when pasting / duplicating frames

- [ ] The management of savestates on which piano roll sheets are based should get several enhancements:
  - [ ] Turning Piano Roll sheets into a properly recorded .m64 movie is a confusing chore - it is currently very difficult to run the correct sheets in the correct order to produce the desired .m64 output, in part due to the carousell button used to cycle between sheets, which will inherently load their respective savestate and run their associated inputs - even when that's not desired.

  - [ ] The savestate from which a sheet is run should be subject to change at the user's will - it should be possible to run the same (semantic, as defined by the `TASState`) inputs from any chosen state. This state may be the end state of another sheet; details need to be worked out.
  
  - [ ] It should be possible to "prepend" frames to a sheet, for example from an existing .m64 movie, because the initial choice for the start of a piano roll sheet may very realistically be insufficient for the task it was meant to solve.
  
  - [ ] For long sheets, it may be beneficial to introduce intermediate savestates to reduce the amount of work done for reruns and consequently improve performance. This needs to be approached with extreme care though, as this is effectively going to be a greedy caching system, which is particularly sensitive to external requirements, and may thus become "buggy" extremely quickly.

- [ ] It would be nice to be able to "base one piano roll sheet ontop of another" - this would most definitely intertwine with the savestate management as mentioned above, but is technically its own enhancement idea.

- [ ] Complex operations across multiple frames may be implemented; for example, a range of frames may be designated to be meant for a "dive-grind", which could be bruteforced via either the lua itself or some external program it communicates with to achieve the result.
The simplest known (and already implemented) "complex operation" is the arctan strain. If this point is implemented in a more generalized manner, then arctan straining should be treated accordingly.

- [ ] Given the above, it is almost inevitable that the `PianoRoll` view needs multiple "tabs" to be presentable - as such, a system must be worked out that manages those tabs as appropriate. There are already components in place that prevent the rendering of anything else in specific, well defined states - this must be streamlined into a UI concept that displays only the information that is relevant to the user's intent at all times.

- [ ] In its current state, the Piano Roll GUI has no options to change its appearance. It may be desirable to define stylers for the parts of it that could (and should) be defined as semantically isolated controls.

- [ ] Auto-saving Piano Rolls would be nice, so that work can be continued after closing mupen without worry - although it's difficult to define _what_ to auto-save and how, at this time.

# Bugs
- [ ] The `Settings` variable, which used to track the state of the `TAS` view, is no longer responsible for the state that tab actually finds itself in anymore. This may be undesirable if Piano rolls aren't used in a particular workflow. (Sorry, but you could just be smart and use piano rolls instead and save and load them like a sane person will be doing in 2038 :finnadie: )

- [ ] A potentially unncessary (or, at the very least poorly placed) `VarWatch_update` call in the `PianoRoll`'s main `at_update_screen` callback causes at least the speed efficiency variable to be wrongly displayed.

- [ ] Not technically a bug, but the `ugui.end_frame()` in the save and load logic invoking the file dialogs should be removed in a future release that may be "shipped" with the according bug fix in mupen that makes the "hack" unnecessary or redundant (idk which of the two it really is)

- [ ] Also not technically a bug, some of the code used to implement the feature is implemented using practices that are questionable at best, and it won't do any harm to improve on the "formal" cleanliness of the code for the Piano Roll along with any APIs it relies on.

**Comment by FramePerfection** [comment]
Once #33 is merged, I will update this issue to reflect the state of the tool by then, as well as elaborating on the several feature ideas that can only exist in the context of the semantic workflow, listed here in short:
- Semi-automatic conversion from .m64 files to semantic workflow sheets
- Dedicated sections for special tasks, e.g. dive grinding
- More elaborate section end conditions (e.g. section end at a speed threshold)
- "Section prebuffers" that allow making input choices n frames _before_ a section starts, using in-memory savestates
- Git-integrations for "undo", "redo", "auto-save" and more
- Brute-forcer integration for sections and/or sheets?!
- Proper usage of icons for GUI elements

As well as some known issues:
- Opening a project file will not play back a section properly, but pause the game after some time (usually the section timeout, I think) anyways.
- Opening a project may unfortunately still cause a crash that seems to be related to the in-memory savestate API. Maybe making creating multiple "temporally overlapping" savestate callbacks is illegal or broken somehow?
- In-memory savestates appear to be uncompressed, which is very bad for git based workflows that may keep many savestates in their history.
